### ISSUE TITLE
Handling of GX2Invalidate modes

### ISSUE THREAD

**Issue by achurch** [decaf-emu/decaf-emu#281]
Since the fix to send proper surfaceSync packets for GX2Invalidate (d95e370), Xenoblade's frame rate has dropped precipitously due to constantly rehashing texture and data buffers, which makes me think the current code on the OpenGL side isn't interpreting the mode bits correctly. In particular, given that the CPU bit triggers a DCFlushRange call, I suspect that we only need to trigger an upload if the CPU bit is set. Based on the current GX2Invalidate implementation, though, the CPU bit is not passed along in the PM4 packet, so we may need to either borrow an unused bit from that packet or add back the custom DecafInvalidate packet. (I do think this can be handled as part of the surfaceSync handler on the GPU thread, since writing to the buffer from the CPU side after GX2Invalidate but before the GPU has had a chance to consume the data is presumably API-invalid, and if any games violate that and still happen to work on real hardware, emulating those correctly may be a lost cause anyway without being able to manage GPU memory directly.)

As a point of reference, Xenoblade hashes about 180MB/frame of textures and 1MB/frame of data buffers (with no actual uploads) during the initial "connecting..." screen. On the title screen, that jumps to almost 500MB/frame of textures (1-2MB uploaded) and 50MB of data buffers (<1MB uploaded).

**Comment by achurch** [comment]
So since I apparently can't get to sleep when I have a puzzle stuck in my mind, I've worked this out, and it looks like the signal we actually need to look at is DCFlushRange itself (and presumably DCStoreRange, if anyone uses it). But since that can be called from any core, we need a separate interface for it after all. I've done this in https://github.com/achurch/decaf-emu/commit/e3b9fda6f06ff80b92237b066968a81ca0bf27bd if you want to take a look; as far as I can tell, there are no graphical differences from current master.

Also with respect to the Xenoblade stats, when I looked at the actual GX2Invalidate ranges, the total invalidated size was only about 10% of the amount of hashing done due to lots of overlapping surface buffers, so it may be worth looking into garbage-collecting at least surfaces sooner rather than later.

**Comment by brett19** [comment]
I indicated this to @exjam as well, but he seemed to think that we should be able to use the 'output' flush part of GX2Invalidate to do the invalidation.  Can you maybe explain why this wouldn't work?

**Comment by achurch** [comment]
I'm not sure what "output" refers to, but games can call DCFlushRange on their own rather than passing the CPU flag to GX2Invalidate (Xenoblade, at least, does this in some cases).

**Comment by achurch** [comment]
I may have spoken too soon: Xenoblade omits both GX2Invalidate and DCFlushRange on some code paths which call GX2InitFetchShaderEx to create a fetch shader. (See https://github.com/achurch/decaf-emu/commit/1d9c97a7a185e046a3bf7735e6c07034f1f17e3c, which fixes some "invalid semantic mapping" errors from Xenoblade caused by reusing the same base address for different fetch shaders.)

That said, given how freely Xenoblade uses GX2Invalidate otherwise, this seems odd. @exjam, can you check whether GX2InitFetchShaderEx from gx2.rpl does GX2Invalidate or DCFlushRange itself? Maybe that's just missing from the decaf implementation.

**Comment by brett19** [comment]
Hey @achurch,
GX2InitFetchShaderEx definitely does not call any invalidation code.  Maybe they are doing something else to indicate that the fetch shader is changed, or perhaps the shader blocks themselves are just not cached?

**Comment by achurch** [comment]
I don't see how the shader blocks can be "not cached" -- they're created using the same store instructions as everything else, so they'll go through the CPU cache, and the GPU won't be able to see the data until it's flushed out of the cache one way or another. (Unless the GPU's shader circuitry can somehow snoop the CPU cache? Seems unlikely, but I dunno.)

To be fair, there is a code path that does call GX2Invalidate(CPU | Shader) after GX2InitFetchShaderEx, so it's possible the GX2Invalidate call was just accidentally omitted on this particular path, and the game happens to run correctly because there's enough cache activity to flush the data out manually. But the same argument could be made in reverse, that the GX2Invalidate call in that one code path was added unnecessarily (out of paranoia, or habit, or not reading the API docs, or what have you), so I still tend to think there's a missing piece here.

**Comment by brett19** [comment]
Due to the fact that the cache gets implicitly flushed regardless of any kind of cache invalidation makes me believe that having our code solely relying on that might be dangerous.  Can you think of any ways to solve this?  Note that the reason we use murmur3 is that its one of the fastest hashing algorithms on x64 👎 .

**Comment by brett19** [comment]
Oh.  Apparently there is a faster one: http://cyan4973.github.io/xxHash/

**Comment by achurch** [comment]
Unfortunately, even with a bandwidth-limited hashing algorithm, there's just no way you can hash 500MB/frame at 60fps on current hardware; even 50MB/frame = 3GB/s is a significant chunk of the total DDR3 bandwidth and is probably not feasible given everything else that has to happen. I think the only solution that's completely general is using host GPU RAM as the emulated physical RAM, so the memory is truly shared as on a real Wii U. Other than that, you're going to have to make one assumption or another about when to revalidate GPU objects.

**Comment by brett19** [comment]
Unfortunately, we aren't even able to use the host GPU ram as real ram due to limitations with the OpenGL API.  I actually recently discussed the idea of using `shared memory` by leveraging the memory allocation freedom that Vulkan provides though.  I actually have a solution to our immediate issue (two actually), but I'm not sure what the performance impact will be.  I'll be looking into it tonight.

**Comment by brett19** [comment]
Actually.  Your probably better to implement this on Linux.  Essentially what we need to do is mprotect the pages containing GPU resources and then when we get a segment violation, we mark the data as being written, and then only replace the "guard" once we actually perform invalidation.  I think the main thing I need your help with is building the violation handler which can mark a page in our own page array as having been written.  I've spent all morning figuring out a way to performantly handle this, I think I'll manage to get it coded for Windows tonight.

**Comment by achurch** [comment]
Good call. I considered the segfault handler approach but I wanted to protect exactly the ranges used by graphics resources, which of course you can't do because page size > 256. But hopefully the additional overhead caused by the coarseness won't be too bad. I guess then it's just an issue of the overhead of the host OS's exception handler itself.

**Comment by brett19** [comment]
Yea.  On Windows they actually expose write tracking as a kernel feature, which significantly reduces the cost of tracking (though the enablement of this is not page level but allocation level which might be an issue).  I'm going to look into it a bit more later, it might be worth you taking a look if Linux has any special memory write tracking stuff.  Though using page guards will probably be what we need to do anyways, since sometimes we need to trigger events on reads as well.

**Comment by brett19** [comment]
Gezus.  This is going to be an extremely complicated project to implement for Windows and Linux.  Going to do this another time, or let someone else handle it.  I put up a branch with the very very very beginnings (decaf-emu/decaf-emu#write-watch)...

**Comment by achurch** [comment]
@brett19
>GX2InitFetchShaderEx definitely does not call any invalidation code.

Sorry, gx2.rpl disagrees with you ;)

**Comment by exjam** [comment]
I see GX2Invalidate(0x40, ...) in GX2InitFetchShaderEx

**Comment by brett19** [comment]
Me too.  Wtf was I smoking yesterday.

**Comment by exjam** [comment]
Can we close this @achurch ? :+1:

**Comment by achurch** [comment]
Was actually just looking at this yesterday and it reminded me to send the DCInvalidate PR. (: (#362) Once that's in, we can close this, yeah.

**Comment by exjam** [comment]
Updated status on this - we can only use this as a performance hint. Not all games actually consistently invalidate memory between CPU and GPU so we cannot solely rely on it.

We can provide it as a configurable performance option for games which are known to accurately invalidate data in which we do not do change detection on the gpu backend such as via mprotect or hashing.
